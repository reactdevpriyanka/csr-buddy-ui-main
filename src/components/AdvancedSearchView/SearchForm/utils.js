import { subWeeks, isValid } from 'date-fns';
import * as Yup from 'yup';

export const phoneRegExp = /^(\+?\d{0,4})?\s?-?\s?(\(?\d{3}\)?)\s?-?\s?(\(?\d{3}\)?)\s?-?\s?(\(?\d{4}\)?)?$/;

export const validationSchema = Yup.object({
  orderId: Yup.string().trim().matches(/^\d*$/, 'Must be numeric characters'),
  status: Yup.string().trim(),
  name: Yup.string().trim(),
  logonId: Yup.string().trim(),
  address: Yup.string().trim(),
  city: Yup.string().trim(),
  zip: Yup.string().matches(
    /^\d{5}(?:-\d{4})?$/,
    'Must be 5 or 9 digits with a format of ##### or #####-####',
  ),
  email: Yup.string().email().trim(),
  phone: Yup.string()
    .trim()
    .matches(phoneRegExp, 'Phone number is not valid')
    .min(7, 'Phone number is not valid')
    .max(16, 'Phone number is not valid'),
  partNumber: Yup.string().trim(),
  blocked: Yup.string().trim(),
  blockReason: Yup.string().trim(),
  fulfillmentCenter: Yup.string().trim(),
  account: Yup.string().trim().min(4).max(4),
  payReferenceId: Yup.string().trim(),
  paypalEmail: Yup.string().email().trim(),
  ipAddress: Yup.string().trim(),
  timePlacedFrom: Yup.string()
    .trim()
    .test('duration', 'The start date must be before the end date', function (value) {
      return checkDuration(value, this.parent.timePlacedTo, this);
    })
    .nullable(),
  timePlacedTo: Yup.string().trim().nullable(),
  timeUpdatedFrom: Yup.string()
    .trim()
    .test('duration', 'The start date must be before the end date', function (value) {
      return checkDuration(value, this.parent.timeUpdatedTo, this);
    })
    .nullable(),
  timeUpdatedTo: Yup.string().trim().nullable(),
  subscriptionId: Yup.string().trim(),
});

export const OrderfilterFormWhiteList = [
  'id',
  'orderId',
  'memberId',
  'address',
  'email',
  'name',
  'phone',
  'ipAddress',
  'logonId',
  'status',
  'statuses',
  'excludedStatuses',
  'siteId',
  'paypalEmail',
  'payReferenceIds',
  'account',
  'donationOrgId',
  'subscriptionId',
  'parentOrderId',
  'blockReason',
  'fulfillmentCenter',
  'partNumber',
];

export function isValidDateRange(startDateVal, endDateVal) {
  if (startDateVal && endDateVal) {
    const startDate = new Date(startDateVal);
    const endDate = new Date(endDateVal);
    if (startDate.getTime() > endDate.getTime()) {
      return false;
    }
    const twoWeeksAgo = subWeeks(endDate, 2);
    if (startDate.getTime() < twoWeeksAgo.getTime()) {
      return false;
    }
  }
  return true;
}

const orderFiltersOnlySearchWithThemNotAllowed = new Set(['city', 'zip']);

/**
 * Returns a list of errors on a filter form if there are any
 *
 * @param {object} validationSchema // The schema used to validate the current form
 * @param {object} values // The current values for the form
 * @returns {object} // The list of errors the form currently has.  If none then the form is currently valid or not dirty
 */
export const doFilterFormValidation = async (
  validationSchema,
  filterFormWhiteList = [],
  values,
) => {
  let errors = {};

  if (!values) {
    return errors;
  }

  const validationOptions = {
    abortEarly: false,
  };

  await validationSchema.validate(values, validationOptions).catch((error) => {
    for (let i = 0; i < error.inner.length; i++) {
      const currentError = error.inner[i];
      const currErrorParams = currentError.params;

      if (currErrorParams.value !== '' && currErrorParams.value !== null) {
        errors[currErrorParams.path] = currentError.message;
      }
    }
  });

  if (JSON.stringify(errors) === '{}' && filterFormWhiteList.length > 0) {
    const results = Object.entries(values).filter(
      ([name, value]) => value !== '' && value !== null,
    );

    if (results.length === 1 && !filterFormWhiteList.includes(results[0][0])) {
      errors[results[0][0]] = 'There must be at least 2 items to search on';
    }

    if (results.length === 3 && orderFiltersOnlySearchWithThemNotAllowed.has(results[0][0])) {
      errors[results[0][0]] =
        'at least one of memberId, address, name, phone, statuses, excludedStatuses, siteId, paypalEmail, payReferenceIds, account, donationOrgId, subscriptionId, parentOrderId, blockReasons, fulfillmentCenters, partNumbers, or trackingId must be set';
    }
  }

  return errors;
};

export function checkDuration(startDateStr, endDateStr, ctx) {
  if (startDateStr && endDateStr) {
    const startDate = new Date(startDateStr);
    const endDate = new Date(endDateStr);
    if (!isValid(startDate) || !isValid(endDate)) {
      return false;
    }
    if (startDate.getTime() > endDate.getTime()) {
      return false;
    }
    const twoWeeksAgo = subWeeks(endDate, 2);
    if (startDate.getTime() < twoWeeksAgo.getTime()) {
      return ctx.createError({ message: 'The date range should be less than 2 weeks' });
    }
  }
  return true;
}
