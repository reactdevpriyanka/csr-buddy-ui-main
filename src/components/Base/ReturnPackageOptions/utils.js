export const LABEL_CHOICES = [
  {
    label: '1 Label',
    enabled: true,
    visible: true,
    value: 1,
  },
  {
    label: '2 Labels',
    enabled: true,
    visible: true,
    value: 2,
  },
  {
    label: '3 Labels',
    enabled: true,
    visible: true,
    value: 3,
  },
  {
    label: '4 Labels',
    enabled: true,
    visible: true,
    value: 4,
  },
  {
    label: '5 Labels',
    enabled: true,
    visible: true,
    value: 5,
  },
  {
    label: '6 Labels',
    enabled: true,
    visible: true,
    value: 6,
  },
];

export const BOX_CHOICES = [
  {
    label: '1 Box',
    enabled: true,
    visible: true,
    value: 1,
  },
  {
    label: '2 Boxes',
    enabled: true,
    visible: true,
    value: 2,
  },
  {
    label: '3 Boxes',
    enabled: true,
    visible: true,
    value: 3,
  },
  {
    label: '4 Boxes',
    enabled: true,
    visible: true,
    value: 4,
  },
  {
    label: '5 Boxes',
    enabled: true,
    visible: true,
    value: 5,
  },
  {
    label: '6 Boxes',
    enabled: true,
    visible: true,
    value: 6,
  },
];
