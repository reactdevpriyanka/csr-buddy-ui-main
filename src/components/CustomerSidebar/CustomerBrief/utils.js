export const isWeight = (name) => name === 'WEIGHT_LIMIT';

export const blacklistedTags = new Set(['DISCOUNT_ABUSER', 'RETURNS_ABUSER', 'SUSPECTED_FRAUD']);
