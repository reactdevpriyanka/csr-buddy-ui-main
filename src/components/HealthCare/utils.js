import { CLOSE_BUTTONS } from '../Base/BaseDialog';

export const closeConsentDialog = (response, setConsentGiven, setDialogOpen) => {
  switch (response) {
    case CLOSE_BUTTONS.CANCEL_BUTTON:
      setConsentGiven(false);
      break;
    case true:
      setConsentGiven(true);
      break;
    default:
      break;
  }
  setDialogOpen(false);
};

export const insurablePetsFilter = (pets = []) => {
  return pets?.filter(
    (pet) =>
      !(
        pet?.petInsurancePolicies?.some(
          (policy) => policy?.productType === 'Insurance' && policy?.policyStatus === 'Active',
        ) &&
        pet?.petInsurancePolicies?.some(
          (policy) => policy?.productType === 'Wellness' && policy?.policyStatus === 'Active',
        )
      ) &&
      (pet?.type?.name === 'Dog' || pet?.type?.name === 'Cat'),
  );
};
