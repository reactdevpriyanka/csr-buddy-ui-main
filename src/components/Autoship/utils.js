import { getDynamicString } from '@/utils/string';
import { formatDate, isCardExpired } from '@utils/dates';

export const renderTitle = ({ name, stat, classes, isDetails = false, getLang }) => {
  let defaultStr = `Autoship "${name}" Details`;

  if (isDetails) {
    return getDynamicString({
      key: 'autoshipCardTitle',
      fallback: defaultStr,
      substitutions: [name],
      getLang,
    });
  }

  if (stat === Status.CANCELLED) {
    defaultStr = `Cancelled Autoship "${name}"`;
    return (
      <span className={classes.cancelledHeaderTitle}>
        {getDynamicString({
          key: 'autoshipCardCancelTitle',
          fallback: defaultStr,
          substitutions: [name],
          getLang,
        })}
      </span>
    );
  } else if (stat === Status.UPCOMING) {
    defaultStr = `Upcoming Shipment for "${name}" Autoship`;
    return getDynamicString({
      key: 'autoshipCardUpcomingTitle',
      fallback: defaultStr,
      substitutions: [name],
      getLang,
    });
  }
  // else { // fix this after BE type data is avilable
  //   defaultStr = `"${name}" Autoship created`;
  //   return getDynamicString({
  //     key: 'autoshipCardCreatedTitle',
  //     fallback: defaultStr,
  //     substitutions: [name],
  //     getLang,
  //   });
  // }
  else {
    defaultStr = `Upcoming Shipment for "${name}" Autoship`;
    return getDynamicString({
      key: 'autoshipCardUpcomingTitle',
      fallback: defaultStr,
      substitutions: [name],
      getLang,
    });
  }
};

export const Status = {
  UPCOMING: Symbol.for('UPCOMING'),
  CANCELLED: Symbol.for('CANCELED'),
  CREATED: Symbol.for('CREATED'),
};

export const getStat = (isUpcoming, status) =>
  isUpcoming ? Status.UPCOMING : status === 'CANCELED' ? Status.CANCELLED : Status.CREATED;

export const makeSubtitle = ({
  date,
  stat,
  lastShipmentDate,
  startDate,
  showCardExpired,
  showLastShipmentDeclined,
  frequency,
  subscriptionId,
  getLang,
}) => {
  let subtitle = [];

  if (frequency) {
    subtitle.push({ text: `Frequency: ${frequency}` });
  }

  let defaultStr = `Cancelled on ${formatDate(date, false)}`;
  let shipmentDefaultStr = `Next shipment on ${formatDate(date, false)}`;

  const nextDateStr =
    stat === Status.CANCELLED
      ? getDynamicString({
          key: 'autoshipNextDate',
          fallback: defaultStr,
          substitutions: [`${formatDate(date, false)}`],
          getLang,
        })
      : getDynamicString({
          key: 'shipNextDate',
          fallback: shipmentDefaultStr,
          substitutions: [`${formatDate(date, false)}`],
          getLang,
        });

  subtitle.push({
    text: nextDateStr,
  });

  // eslint-disable-next-line unicorn/no-array-push-push
  subtitle.push({ text: `Last shipment on ${formatDate(lastShipmentDate, false)}` });

  if (showCardExpired) {
    subtitle.push({
      text: 'Upcoming shipment will not ship out due to expired card.',
      isError: true,
    });
  }

  if (showLastShipmentDeclined) {
    subtitle.push({
      text: 'Last shipment was declined due to expired card.',
      isError: true,
    });
  }

  if (subscriptionId) {
    subtitle.push({
      text: `Autoship ID: ${subscriptionId}`,
    });
  }

  return subtitle;
};

export const showCardExpired = (subscriptionData) => {
  return subscriptionData?.paymentMethods?.some(({ type, card }) => {
    // Skip showing "expired card" message if PayPal:
    if (type === 'PAYPAL' || !card?.expirationMonth || !card?.expirationYear) {
      return false;
    }
    const { expirationMonth, expirationYear } = card;
    return isCardExpired(expirationMonth, expirationYear);
  });
};

export const getPromotions = (subscriptionData) => {
  return subscriptionData.promotions.reduce((result, item) => {
    const tmpAdjments = item.adjustments.filter((item) => item.type === 'PROMOTION');

    for (const adjustment of tmpAdjments) {
      const obj = result[adjustment.code];

      if (obj) {
        obj.value += adjustment.amount;
      } else {
        result[adjustment.code] = {
          promotion: adjustment.description,
          value: adjustment.amount,
          code: adjustment.code,
        };
      }
    }

    return result;
  }, {});
};
