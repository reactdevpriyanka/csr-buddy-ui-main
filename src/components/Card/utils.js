import { snakeCaseToTitleCase } from '@/utils/string';

const validReturnStates = new Set([
  'SUBMITTED',
  'VERIFYING',
  'RECEIVE_WAIT',
  'PAY_WAIT',
  'GIFT_CARD_CASH_OUT_WAIT',
  'GIFT_CARD_CLOSE_WAIT',
  'GIFT_CARD_CLOSED',
  'VERIFIED',
  'PROCESSING',
  'TAX_WAIT',
  'PROCESSED',
  'CANCELLING',
  'CANCELING',
  'CANCELLED',
  'CANCELED',
  'IMPORTED_IN_FLIGHT',
  'VET_DIET',
  'FAILED',
]);

export const ITEM_TYPE = {
  DISCONTINUED: Symbol.for('DISCONTINUED'),
  OUT_OF_STOCK: Symbol.for('OUT_OF_STOCK'),
  PERSONALIZED: Symbol.for('PERSONALIZED'),
  FROZEN: Symbol.for('FROZEN'),
  CONNECT_WITH_A_VET: Symbol.for('CONNECT_WITH_A_VET'),
  PHARMACY: Symbol.for('PHARMACY'),
  DROPSHIP: Symbol.for('DROPSHIP'),
  FORCED_BACKORDER: Symbol.for('FORCED_BACK_ORDER'),
  BUNDLED_ITEM: Symbol.for('BUNDLED'),
  VET_DIET: Symbol.for('VET_DIET'),
};

export const ITEM_TYPES_LABELS = {
  [ITEM_TYPE.DISCONTINUED]: 'DISCONTINUED ITEM',
  [ITEM_TYPE.OUT_OF_STOCK]: 'OUT OF STOCK',
  [ITEM_TYPE.PERSONALIZED]: 'PERSONALIZED ITEM',
  [ITEM_TYPE.CONNECT_WITH_A_VET]: 'CONNECT WITH A VET',
  [ITEM_TYPE.PHARMACY]: 'PHARMACY ITEM',
  [ITEM_TYPE.DROPSHIP]: 'DROPSHIPPED ITEM',
  [ITEM_TYPE.FORCED_BACKORDER]: 'FORCED BACKORDER',
  [ITEM_TYPE.BUNDLED_ITEM]: 'BUNDLED ITEM',
  [ITEM_TYPE.VET_DIET]: 'VET DIET ITEM',
};

export const formatItemType = (status) => {
  return ITEM_TYPES_LABELS[Symbol.for(status)] || status;
};

export const notInStock = (tag) => {
  return tag.includes('OUT_OF_STOCK') || tag.includes('FORCED_BACK_ORDER');
};

export const formatReturnType = (enumStr) => {
  switch (enumStr) {
    case 'PRODUCT_CONCESSION':
      return 'Concession';
    default:
      return snakeCaseToTitleCase(enumStr);
  }
};

export const formatReturnState = (state) => {
  switch (state) {
    case 'FAILED':
      return 'Failed';
    case 'CANCELED':
    case 'CANCELLED':
      return 'Canceled';
    case 'PROCESSED':
      return 'Completed';
    default:
      return validReturnStates.has(state?.toUpperCase()) ? 'Pending' : state;
  }
};

export const formatAction = (type, state) =>
  `${formatReturnType(type)} ${formatReturnState(state)}`;
