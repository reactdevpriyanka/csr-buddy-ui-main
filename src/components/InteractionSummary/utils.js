import PhoneCallbackIcon from '@material-ui/icons/PhoneCallback';
import ForumIcon from '@material-ui/icons/Forum';
import EmailIcon from '@material-ui/icons/Email';
import { snakeCaseToTitleCase } from '@/utils/string';
import { useMemo } from 'react';
import _groupBy from 'lodash/groupBy';

export const getContactReason = (details) => {
  if (details?.tertiary) return snakeCaseToTitleCase(details.tertiary);
  else if (details?.secondary) return snakeCaseToTitleCase(details.secondary);
  else if (details?.primary) return snakeCaseToTitleCase(details.primary);
  else return null;
};

export const getContactChannelIcon = (channel) => {
  switch ((channel || '').toLowerCase()) {
    case 'chat':
      return <ForumIcon />;
    case 'email':
    case 'online form':
      return <EmailIcon />;
    default:
      return <PhoneCallbackIcon />;
  }
};

export const useShippingAppeasements = (appeasements) => {
  return useMemo(() => {
    let includeShippingConcession = appeasements.some(
      (appeasement) => appeasement?.description === 'shipping',
    );

    if (!includeShippingConcession) {
      return appeasements;
    }

    let mergedAppeasements = [];
    let shippingAppeasements = appeasements.filter(
      ({ actions }) => actions?.[0]?.type === 'SHIPPING_CONCESSION',
    );
    let shippingAppeasementsGroupedById = _groupBy(
      shippingAppeasements,
      (appeasement) => appeasement.appeasementId,
    );

    // filter out shipping concession appeasements before we push merged shipping appeasements in
    mergedAppeasements = appeasements.filter(
      ({ actions }) => actions?.[0]?.type !== 'SHIPPING_CONCESSION',
    );

    // Merge the shipping concessions (dummy shipping line item is added in ConcessionTable on submit)
    for (const appeasementId in shippingAppeasementsGroupedById) {
      let mergedShippingAppeasement = shippingAppeasementsGroupedById[appeasementId].reduce(
        (merged, currAppeasement) => {
          return { ...currAppeasement, ...merged };
        },
        {},
      );

      mergedAppeasements.push(mergedShippingAppeasement);
    }

    return mergedAppeasements;
  }, [appeasements]);
};
