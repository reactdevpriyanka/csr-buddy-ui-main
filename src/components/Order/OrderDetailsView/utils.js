import { getDayDateYearTimeTimezone } from '@/utils';
import { multiply } from 'lodash';

export const AllowableActions = {
  RETURN_ITEMS: 'RETURN_ITEMS',
  SEND_ORDER_INVOICE: 'SEND_ORDER_INVOICE',
  SEND_ORDER_CONFIRMATION: 'SEND_ORDER_CONFIRMATION',
  SEND_ORDER_CANCELED_NOTIFICATION: 'SEND_ORDER_CANCELED_NOTIFICATION',
  ALLOCATE_INVENTORY: 'ALLOCATE_INVENTORY',
  ADD_FULFILLMENT_COMMENT: 'ADD_FULFILLMENT_COMMENT',
  BLOCK_ORDER: 'BLOCK_ORDER',
  FORCE_RELEASE_ORDER: 'FORCE_RELEASE_ORDER',
  PROCESS_ORDER: 'PROCESS_ORDER',
  CANCEL_ORDER: 'CANCEL_ORDER',
  EDIT_SHIPPING_ADDRESS: 'EDIT_SHIPPING_ADDRESS',
  ADD_PROMOTION_CODE: 'ADD_PROMOTION_CODE',
  ADD_PRICE_ADJUSTMENT: 'ADD_PRICE_ADJUSTMENT',
  REMOVE_ITEM: 'REMOVE_ITEM',
  REDUCE_QUANTITY: 'REDUCE_QUANTITY',
  CREATE_PACKAGE: 'CREATE_PACKAGE',
  CONFIRM_SHIPMENT: 'CONFIRM_SHIPMENT',
  RESEND_SHIPMENT_MESSAGE: 'RESEND_SHIPMENT_MESSAGE',
  FORCE_CANCEL_RELEASE: 'FORCE_CANCEL_RELEASE',
  CANCEL_RELEASE: 'CANCEL_RELEASE',
  CANCEL_RETURN: 'CANCEL_RETURN',
  MARK_AS_RECEIVED: 'MARK_AS_RECEIVED',
  CREATE_NEW_LABELS: 'CREATE_NEW_LABELS',
};

export const AllowableActionTypes = {
  ORDER_ACTIONS: 'orderActions',
  RELEASE_ACTIONS: 'releaseActions',
  RETURN_ACTIONS: 'returnActions',
};

export const PackageReleaseTypes = Object.freeze({
  PACKAGE: Symbol('package'),
  RELEASE: Symbol('release'),
});

// format supplemental data attributes to for tooltip text
export const getAttributes = (tag, attributesMap, vetContactInfo, product, rowData, testId) => {
  switch (tag) {
    case 'DISCONTINUED':
      return (
        <span>
          {product?.discontinueDate != null && (
            <span data-testid="tag:discontinueDate" key="discontinueDate">
              <b>Discontinued Date:</b> {getDayDateYearTimeTimezone(product.discontinueDate)}
            </span>
          )}
        </span>
      );
    // case 'OUT_OF_STOCK': Disabling this until Force BackOrder and OOS back in stock dates are available in the api
    //   return (
    //     <span>
    //       {product?.discontinueDate != null && (
    //         <span key="discontinueDate">
    //           <b>Back in Stock Date:</b>
    //         </span>
    //       )}
    //     </span>
    //   );
    // case 'FORCED_BACK_ORDER':
    //   return (
    //     <span>
    //       {product?.discontinueDate != null && (
    //         <span key="discontinueDate">
    //           <b>Available Date:</b>
    //         </span>
    //       )}
    //     </span>
    //   );
    case 'CONNECT_WITH_A_VET':
      return attributesMap ? (
        <span>
          {vetContactInfo?.petName != null && (
            <span data-testid="tag:petName" key="petName">
              Pet Name: {vetContactInfo.petName}
              <br />
            </span>
          )}
          {attributesMap?.reason != null && (
            <span data-testid="tag:appointmentReason" key="appointmentReason">
              Appointment Reason: {attributesMap.reason}
              <br />
            </span>
          )}
          {attributesMap?.id != null && (
            <span data-testid="tag:appointmentId" key="appointmentId">
              Appointment ID: {attributesMap.id}
              <br />
            </span>
          )}
          {attributesMap?.endTime != null && (
            <span data-testid="tag:appointmentEndTime" key="appointmentEndTime">
              Appointment End Time: {`${getDayDateYearTimeTimezone(attributesMap.endTime)}`}
              <br />
            </span>
          )}
        </span>
      ) : null;
    case 'PHARMACY':
      return vetContactInfo ? (
        <span>
          {vetContactInfo?.clinicName != null && (
            <span data-testid="tag:clinicName" key="clinicName">
              Clinic Name: {vetContactInfo.clinicName}
              <br />
            </span>
          )}
          {vetContactInfo?.contactVet != null && (
            <span data-testid="tag:vetContacted" key="vetContacted">
              Vet Contacted: {vetContactInfo.contactVet ? 'Yes' : 'No'}
              <br />
            </span>
          )}
          {vetContactInfo?.vetProfileId != null && (
            <span data-testid="tag:vetProfileId" key="vetProfileId">
              Vet Profile ID: {vetContactInfo.vetProfileId}
              <br />
            </span>
          )}
          {vetContactInfo?.petName != null && (
            <span data-testid="tag:petName" key="petName">
              Pet Name: {vetContactInfo.petName}
              <br />
            </span>
          )}
        </span>
      ) : null;
    case 'DROPSHIP':
      return (
        <span>
          {rowData?.minDeliveryDate != null && rowData?.maxDeliveryDate != null && (
            <span data-testid="tag:minDeliveryDate" key="minDeliveryDate">
              <b>Delivers between:</b> {getDayDateYearTimeTimezone(rowData.minDeliveryDate)} and{' '}
              {getDayDateYearTimeTimezone(rowData.maxDeliveryDate)}
            </span>
          )}
        </span>
      );
    case 'PERSONALIZED':
      const entries = Object.entries(product?.personalizationAttributeMap || {}).sort(
        (obj1, obj2) => {
          return obj1[0].localeCompare(obj2[0]);
        },
      );
      return (
        <span>
          {entries?.map((entry) => {
            return (
              <div key={entry[0]}>
                <span data-testid={`tag:personalized:${entry[0]}`}>
                  <b>{entry[0]}:</b> {entry[1]}
                </span>
              </div>
            );
          })}
        </span>
      );
    default:
      return null;
  }
};

export const generateReturnItemKey = (returnItem) => returnItem.type + '_' + returnItem.state;

// Sums up the total quantity returned for each type of unique return
export const createTotalReturnedQuantities = (returnItems) => {
  const totalReturnedQuantities = {};

  for (const returnItem of returnItems) {
    const key = returnItem.lineItemId;
    const innerKey = generateReturnItemKey(returnItem);

    if (!totalReturnedQuantities[key]) {
      totalReturnedQuantities[key] = {};
    }

    if (!totalReturnedQuantities[key][innerKey]) {
      totalReturnedQuantities[key][innerKey] = 0;
    }

    // When either type of concession fails, the return service gives us back a totalCredit of 0 and no quantity field.
    // Because of this, we treat each concession like a totalCredit of 1 so it can be distributed to shipments like other concessions
    let specialCaseIncrement = 0;
    if (innerKey === 'PRODUCT_CONCESSION_FAILED' || innerKey === 'SHIPPING_CONCESSION_FAILED') {
      specialCaseIncrement = 1;
    }

    const increment =
      returnItem.type === 'PRODUCT_CONCESSION' || returnItem.type === 'SHIPPING_CONCESSION'
        ? returnItem?.totalCredit + specialCaseIncrement
        : returnItem?.quantity + specialCaseIncrement;

    totalReturnedQuantities[key][innerKey] += increment;
  }

  return totalReturnedQuantities;
};

export const orderHasItemsSpreadAcrossShipments = (shipments, currShipment, itemMap = {}) => {
  if (shipments?.length === 1) return false;

  const lineItemIds = Object.keys(itemMap);
  for (const lineItemId of lineItemIds) {
    const otherShipmentLineItemIds = shipments
      ?.filter((shipment) => shipment.id !== currShipment.id)
      ?.map((shipment) => shipment?.shipmentItems)
      ?.flat()
      ?.map(({ lineItemId }) => lineItemId);
    if (otherShipmentLineItemIds?.includes(lineItemId)) return true;
  }
  return false;
};

// When the same lineItemId is found across multiple shipments, we have no way of knowing which returnItems (if any)
// are attributed to any specific item in any specific shipment.
//
// The number of returns processed for a lineItemId will always be less than the total quantity of that lineItemId across all shipments
// so we iterate through the shipments untill we run out of return tags for that lineItemId.
export const assignReturnsToShipments = (shipments, returnItems, itemMap = {}) => {
  const totalReturnedQuantities = createTotalReturnedQuantities(returnItems);

  for (const [index, shipment] of shipments.entries()) {
    for (const shipmentItem of shipment.shipmentItems) {
      const lineItemId = shipmentItem.lineItemId;

      if (totalReturnedQuantities[lineItemId]) {
        const returnTagsToRender = Object.entries(totalReturnedQuantities[lineItemId])
          ?.filter(([returnItemKey, quantityReturned]) => quantityReturned >= 0.001)
          ?.map(([returnItemKey, quantityReturned]) => returnItemKey);

        shipments[index] = {
          ...shipment,
          returnTagsToRender: returnTagsToRender,
        };

        for (const returnItemKey in totalReturnedQuantities[lineItemId]) {
          const returnedQuantity = totalReturnedQuantities[lineItemId][returnItemKey];
          if (returnedQuantity < 0) continue; // No need to continue to compute if there is no returnable quantity left for a lineItemId

          const returnItem = returnItems.find(
            (item) => generateReturnItemKey(item) === returnItemKey,
          );

          if (!returnItem) continue;

          let quantityToDecrement;

          switch (returnItem?.type) {
            case 'PRODUCT_CONCESSION':
              quantityToDecrement = multiply(
                shipmentItem?.quantity * Number.parseFloat(itemMap[lineItemId]?.unitPrice?.value),
              );
              break;
            case 'SHIPPING_CONCESSION':
              quantityToDecrement = returnItem?.totalCredit;
              break;
            case 'REPLACEMENT':
            case 'REFUND':
              quantityToDecrement = shipmentItem?.quantity;
              break;
            default:
              quantityToDecrement = 0;
              break;
          }

          // Subtract the decrement quantity from the quantity of items (or total value of products in concession scenarios)
          // returned to only add returns to be rendered to a shipment if there were enough returns to justify it
          // when the same lineItemId is spread across multiple packages
          totalReturnedQuantities[lineItemId][returnItemKey] =
            returnedQuantity - quantityToDecrement;
        }
      }
    }
  }
};
