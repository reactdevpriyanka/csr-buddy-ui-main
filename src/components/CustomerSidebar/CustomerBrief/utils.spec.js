import { isWeight, blacklistedTags } from './utils';

describe('Tag/utils', () => {
  describe('isWeight', () => {
    test('it should return true if tag name is WEIGHT_LIMIT', () => {
      expect(isWeight('WEIGHT_LIMIT')).toEqual(true);
    });

    test('it should return false for all else', () => {
      expect(isWeight('NOT_WEIGHT_LIMIT')).toEqual(false);
    });
  });

  describe('blacklistedTags', () => {
    test('it should match snapshot', () => {
      expect(blacklistedTags).toMatchSnapshot();
    });
  });
});
