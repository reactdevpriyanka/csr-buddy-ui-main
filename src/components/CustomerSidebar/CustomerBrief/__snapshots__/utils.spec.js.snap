// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag/utils blacklistedTags it should match snapshot 1`] = `
Set {
  "DISCOUNT_ABUSER",
  "RETURNS_ABUSER",
  "SUSPECTED_FRAUD",
}
`;
